20220831-add-type-restrictions-to-json-syntax.md

```diff
@@ -0,0 +1,558 @@
+- having the following `"member": { "directly_related_user_types": [{ "type": "user", "any": true }, { "type": "employee", "any": true }] }` in the group type definition indicates that the `*` syntax is allowed and that when present, it means all objects of type `user` or `employee` can be directly related to a `group` as `member`
+
+This will affect only relations that are [directly related](https://openfga.dev/docs/modeling/building-blocks/direct-relationships) (as in they are considered "assignable" and have [the direct relationship keyword ("this")](https://openfga.dev/docs/configuration-language#the-direct-relationship-keyword) in their relation definition).
+For relation definitions that:
+
+- have the direct relationship keyword (this): `directly_related_user_types` must be present and MUST be an non-empty array containing at least one type/relation combination
+- do not have the direct relationship keyword: `directly_related_user_types` can optionally be present, but if it does MUST be an empty array
+
+> Note: This syntax does not offer a way to enforce restrictions based on what the userset of group members resolves to (for example, group member can be a user, an employee or another userset). Later on, tooling can help visually indicate this to the user inline.
+
+To prevent breaking changes, this will be done by appending a new field called `metadata` -> `relations` to each type definition. Each relation will be a key under this that is a map that contains a field called `directly_related_user_types` that contains what type or type and relation combination could be directly related to it. If the `*` element is valid for that relation, then the type and `any: true` need to be in the `directly_related_user_types`. If multiple entries that contains `any: true` are present, a `*` in a relationship tuple will be interpreted to mean any user of the union of those types.
+
+A directly related user type cannot contain both a `relation` key and an `any` key. If the `any` key is excluded, it will be interpreted to be `any: false`.
+
+The introduced `metadata` entry in the type definition would look like:
+
+```typescript
+type DirectlyRelatedUserTypeBase = {
+  type: string;
+};
+
+type DirectlyRelatedUserTypeRelation = {
+  relation?: string;
+};
+
+type DirectlyRelatedUserTypeAny = {
+  any?: boolean;
+};
+
+type DirectlyRelatedUserType = DirectlyRelatedUserTypeBase &
+  (DirectlyRelatedUserTypeRelation | DirectlyRelatedUserTypeAny);
+
+type RelationMetadata = {
+  directly_related_user_types?: DirectlyRelatedUserType[];
+};
+
+type Metadata = {
+  relations?: Record<string, RelationMetadata>;
+};
+```
+
+If a tuple such as `user=team:1#member, relation=member, object=group:2` needs to be added, then the `member` relation on the `group` type must have `{"type":"group","relation":"member"}` as one of its directly related user types.
+
+As an example:
+
+```json
+{ "type_definitions": [
+  { "type": "user" },
```

Comment:
https://github.com/openfga/syntax-transformer/blob/main/src/friendly-to-api.ts#L63

This would actually be `type: user, relations: {}`